crates/ursa-pod/benches/e2e.rs

```diff
@@ -0,0 +1,271 @@
+            benchmark_sizes(
+                &mut g,
+                files,
+                unit,
+                http_hyper::client_loop,
+                http_hyper::server_loop,
+            );
+        }
+    }
+}
+
+mod tcp_ufdp {
+    use futures::future::join_all;
+    use tokio::{
+        net::{TcpListener, TcpStream},
+        task,
+    };
+    use ursa_pod::{
+        client::UfdpClient,
+        connection::consts::MAX_BLOCK_SIZE,
+        server::{Backend, UfdpHandler},
+        types::{Blake3Cid, BlsSignature, Secp256k1PublicKey},
+    };
+
+    const DECRYPTION_KEY: [u8; 33] = [3u8; 33];
+    const CLIENT_PUB_KEY: [u8; 48] = [3u8; 48];
+    const CID: Blake3Cid = Blake3Cid([3u8; 32]);
+
+    #[derive(Clone, Copy)]
+    struct DummyBackend {
+        content: &'static [u8],
+    }
+
+    impl Backend for DummyBackend {
+        fn raw_block(&self, _cid: &Blake3Cid, block: u64) -> Option<&[u8]> {
+            let s = block as usize * MAX_BLOCK_SIZE;
+            if s < self.content.len() {
+                let e = self.content.len().min(s + MAX_BLOCK_SIZE);
+                Some(&self.content[s..e])
+            } else {
+                None
+            }
+        }
+
+        fn decryption_key(&self, _request_id: u64) -> (ursa_pod::types::Secp256k1AffinePoint, u64) {
+            (DECRYPTION_KEY, 0)
+        }
+
+        fn get_balance(&self, _pubkey: Secp256k1PublicKey) -> u128 {
+            9001
+        }
+
+        fn save_batch(&self, _batch: BlsSignature) -> Result<(), String> {
+            Ok(())
+        }
+    }
+
+    /// Simple tcp server loop that replies with static content
+    pub async fn server_loop(
+        addr: String,
+        content: &'static [u8],
+        tx_started: tokio::sync::oneshot::Sender<u16>,
+    ) {
+        let listener = TcpListener::bind(&addr).await.unwrap();
+        let port = listener.local_addr().unwrap().port();
+
+        tx_started.send(port).unwrap();
+
+        loop {
+            let (stream, _) = listener.accept().await.unwrap();
+            let handler = UfdpHandler::new(stream, DummyBackend { content }, 0);
+            task::spawn(async move {
+                if let Err(e) = handler.serve().await {
+                    println!("server error: {e:?}");
+                }
+            });
+        }
+    }
+
+    /// Simple client loop that sends a request and loops over the block stream, dropping the bytes
+    /// immediately.
+    pub async fn client_loop(addr: String, iterations: usize) {
+        let mut tasks = vec![];
+        for _ in 0..iterations {
+            let stream = TcpStream::connect(&addr).await.unwrap();
+            let task = task::spawn(async {
+                let mut client = UfdpClient::new(stream, CLIENT_PUB_KEY, None).await.unwrap();
+
+                client.request(CID).await.unwrap();
+            });
+            tasks.push(task);
+        }
+        join_all(tasks).await;
+    }
+}
+
+#[cfg(feature = "bench-hyper")]
+mod http_hyper {
```

Comment:
Why isn't this test using the thing it's supposed to bench mark?

Reply:
What do you mean? It's using hyper's http1 server and service fn that responds with arbitrary static bytes, and a hyper client. This is using hyper 1.0.0-rc3

```
use http_body_util::{BodyExt, Empty, Full};
use hyper::{server::conn::http1, service::service_fn, Request, Response};
```

Reply:
Is the idea here to compare TCP+HTTP vs UFDP?

Reply:
Yes, purely a comparison since the benchmarks are local and vary slightly from machine to machine, it's why it's optional and behind a feature flag